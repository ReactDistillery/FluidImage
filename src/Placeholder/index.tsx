import * as React from "react";

export interface PlaceholderProps {
  crop: Array<number>;
  backgroundColor?: string;
  fillContainer?: boolean;
  placeholderRef?(ref: Element | null): void;
  children?: any;
}

const styles: React.CSSProperties = {
  padding: {
    display: "inline-block",
    width: "100%",
    position: "relative",
    overflow: "hidden",
    lineHeight: 0
  }
};

const generatePlaceholderStyles = (
  crop: Array<number>,
  fillContainer = false,
  backgroundColor = "#f9f9f9"
) => ({
  ...styles.padding,
  backgroundColor: backgroundColor,
  paddingBottom: fillContainer ? 0 : `${100 * (crop[3] / crop[2])}%`,
  height: fillContainer ? "100%" : 0
});

export default class Placeholder extends React.Component<PlaceholderProps, {}> {
  render() {
    const {
      backgroundColor,
      fillContainer,
      crop,
      placeholderRef,
      children
    } = this.props;
    return (
      <span
        ref={placeholderRef}
        style={generatePlaceholderStyles(crop, fillContainer, backgroundColor)}
      >
        {children}
      </span>
    );
  }
}
