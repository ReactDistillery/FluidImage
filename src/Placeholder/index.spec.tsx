import React from "react";
import { create } from "react-test-renderer";
import Placeholder from "./";

const minimalProps = {
  crop: [0, 0, 100, 100]
};

const fullProps = {
  crop: [0, 0, 100, 100],
  color: "red",
  fillContainer: true,
  placeholderRef: <div />
};

describe("FluidImage - Placeholder", () => {
  test("it renders a span with minimal props", () => {
    const placeholder = create(<Placeholder {...minimalProps} />).toJSON();
    expect(placeholder).toMatchSnapshot();
  });
  test("it renders a span with full props", () => {
    const placeholder = create(
      <Placeholder {...minimalProps}>{"children"}</Placeholder>
    ).toJSON();
    expect(placeholder).toMatchSnapshot();
  });
});
