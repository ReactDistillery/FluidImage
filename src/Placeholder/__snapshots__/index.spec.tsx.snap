// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FluidImage - Placeholder it renders a span with full props 1`] = `
<span
  style={
    Object {
      "backgroundColor": "#f9f9f9",
      "display": "inline-block",
      "height": 0,
      "lineHeight": 0,
      "overflow": "hidden",
      "paddingBottom": "100%",
      "position": "relative",
      "width": "100%",
    }
  }
>
  children
</span>
`;

exports[`FluidImage - Placeholder it renders a span with minimal props 1`] = `
<span
  style={
    Object {
      "backgroundColor": "#f9f9f9",
      "display": "inline-block",
      "height": 0,
      "lineHeight": 0,
      "overflow": "hidden",
      "paddingBottom": "100%",
      "position": "relative",
      "width": "100%",
    }
  }
/>
`;
